### PR TITLE
upx: 3.96 -> 4.0.1

### PR DESCRIPTION
**Pull request by LeSuisse**
###### Description of changes

https://github.com/upx/upx/blob/v4.0.1/NEWS

Fixes CVE-2021-30500.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/compression/upx/default.nix [hunk 1]
```diff
@@ -1,31 +1,14 @@
-{ lib, stdenv, fetchurl, ucl, zlib, perl }:
+{ lib, stdenv, fetchurl, cmake }:
 
 stdenv.mkDerivation rec {
   pname = "upx";
-  version = "3.96";
+  version = "4.0.1";
   src = fetchurl {
     url = "https://github.com/upx/upx/releases/download/v${version}/${pname}-${version}-src.tar.xz";
```
Comment (SuperSandro2000) [inline pkgs/tools/compression/upx/default.nix:7]: Is there any difference between this and the source from github?

Comment (LeSuisse) [reply pkgs/tools/compression/upx/default.nix]: I updated the PR to use `fetchFromGitHub`.

**Comment by trofi** [comment]
Result of `nixpkgs-review pr 202254` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>teensyduino</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>brogue</li>
    <li>libtcod</li>
    <li>retdec</li>
    <li>retdec-full</li>
    <li>upx</li>
    <li>vlang</li>
  </ul>
</details>

**Comment by trofi** [comment]
> teensyduino

Looks like at least for this package `upx` produces something undecompressable, can you have a  look?

```
BUILD SUCCESSFUL
Total time: 24 seconds
buildPhase completed in 35 seconds
@nix { "action": "setPhase", "phase": "glibPreInstallPhase" }
glibPreInstallPhase
@nix { "action": "setPhase", "phase": "installPhase" }
installing
                       Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2022
UPX 4.0.1       Markus Oberhumer, Laszlo Molnar & John Reiser   Nov 16th 2022

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
upx: ./TeensyduinoInstall.linux64: Exception: compressed data violation

Unpacked 1 file: 0 ok, 1 error.
```

**Comment by LeSuisse** [comment]
This appears to be a known upstream bug https://github.com/upx/upx/issues/599

**Pull request closed without merging**